NServiceBus.FileBasedRouting/XmlRoutingFile.cs

```diff
@@ -2,42 +2,83 @@
+        public XmlRoutingFile(Func<Stream> streamAccessor)
         {
-            this.filePath = filePath;
+            this.streamAccessor = streamAccessor;
+
+            using (var stream = GetType().Assembly.GetManifestResourceStream("NServiceBus.FileBasedRouting.routing.xsd"))
+            using (var xmlReader = XmlReader.Create(stream))
+            {
+                schema = new XmlSchemaSet();
+                schema.Add("", xmlReader);
+            }
         }
 
         public IEnumerable<EndpointRoutingConfiguration> Read()
         {
-            using (var fileStream = File.OpenRead(filePath))
+            using (var stream = streamAccessor())
             {
-                var document = XDocument.Load(fileStream);
+                var document = XDocument.Load(stream);
+                document.Validate(schema, null, true);
+
                 var endpointElements = document.Root.Descendants("endpoint");
 
                 var configs = new List<EndpointRoutingConfiguration>();
                 foreach (var endpointElement in endpointElements)
                 {
-                    var config = new EndpointRoutingConfiguration();
-                    config.LogicalEndpointName = endpointElement.Attribute("name").Value;
+                    var config = new EndpointRoutingConfiguration
+                    {
+                        LogicalEndpointName = endpointElement.Attribute("name").Value
+                    };
 
-                    config.Commands = endpointElement.Element("handles")
-                        ?.Elements("command")
-                        .Select(e => Type.GetType(e.Attribute("type").Value, true))
-                        .ToArray() ?? new Type[0];
+                    var handles = endpointElement.Element("handles");
 
+                    var separatelyConfiguredCommands = handles
+                                             ?.Elements("command")
+                                             .Select(e => SelectCommand(e.Attribute("type").Value))
+                                             .ToArray() ?? Type.EmptyTypes;
+
+                    var filteredCommands = handles?.Elements("commands").SelectMany(SelectCommands) ?? Type.EmptyTypes;
+
+                    config.Commands = separatelyConfiguredCommands.Concat(filteredCommands).Distinct().ToArray();
                     configs.Add(config);
                 }
 
                 return configs;
             }
         }
+
+        static Type SelectCommand(string typeName)
+        {
+            return Type.GetType(typeName, true);
+        }
+
+        static IEnumerable<Type> SelectCommands(XElement commandsElement)
+        {
+            var assemblyName = commandsElement.Attribute("assembly").Value;
+            var assembly = Assembly.Load(assemblyName);
+            var exportedTypes = assembly.ExportedTypes;
+            var @namespace = commandsElement.Attribute("namespace");
+            if (@namespace == null)
```

Comment:
this is a tricky one. I think if you declare a type without namespace, the namespace is null. But you can't specify null as a namespace attribute value in case you want to select messages with an empty namespace only. Not sure whether we need to address this right now oder if we can raise a new issue for that.